src/Microsoft.AspNetCore.Routing/Matchers/DfaMatcherFactory.cs

```diff
@@ -0,0 +1,38 @@
+﻿// Copyright (c) .NET Foundation. All rights reserved.
+// Licensed under the Apache License, Version 2.0. See License.txt in the project root for license information.
+
+using System;
+using Microsoft.Extensions.DependencyInjection;
+
+namespace Microsoft.AspNetCore.Routing.Matchers
+{
+    internal class DfaMatcherFactory : MatcherFactory
+    {
+        private readonly IServiceProvider _services;
+
+        // Using the service provider here so we can avoid coupling to the dependencies
+        // of DfaMatcherBuilder.
+        public DfaMatcherFactory(IServiceProvider services)
+        {
+            if (services == null)
+            {
+                throw new ArgumentNullException(nameof(services));
+            }
+
+            _services = services;
+        }
+
+        public override Matcher CreateMatcher(EndpointDataSource dataSource)
+        {
+            if (dataSource == null)
+            {
+                throw new ArgumentNullException(nameof(dataSource));
+            }
+
+            return new DataSourceDependentMatcher(dataSource, () =>
+            {
+                return _services.GetRequiredService<DfaMatcherBuilder>();
+            });
```

Comment:
I'm keeping this general plumbing free of the dependencies of the `DfaMatcher`. That will make it less impactful when we add/change stuff since this and `DfaMatcherBuilder` will be used via DI in the perf tests and integration tests.

Reply:
I'm ok with it since it is all internal.